### PR TITLE
feat: references table view actions

### LINKED ISSUE #201: References "Table view" actions
> ...full table with add/remove columns, sort/filter/tag/organize, bulk modifications, etc

From #159 

- [x] Find table component library that supports the features listed above
- [ ] add/remove columns
- [x] sort and filter columns
- [x] quick filter
- [x] Allow the references table view to open (refresh) with a predefined filter (add "link" to author name in the references details view)
- [x] bulk modifications

### PR DESCRIPTION
**Pull request by cguedes**
This adds a data grid to view references with:

- filtering (global, with any text)
- column sort
- column filter (hover on header and use widget)
- column order
- resizable coluns
- pinned column (first with cite key)
- custom cell render (for status)
- selectable rows (SPACE or click in checkbox)
- multiple rows selection (with range selection)
- persistent column state (position, sort, visibility)
- top action buttons
    - add new reference
    - bulk remove (only UI - implementation tbd in #156)
    - bulk export (only UI - implementation tbd in #214)
    - reset action for the column state
- click on an author in References List (left panel) will open/update the references table with filter
- initial support to edit references data (ex: title)
    - one can double-click in the title cell, but for now only displays action in the console

image

image

This closes #201

### CODE REVIEW

File: src/application/components/MainPanel.tsx [hunk 1]
```diff
@@ -59,6 +59,7 @@ interface MainPanelPaneProps {
 export function MainPanelPane({ pane, pdfViewerRef }: MainPanelPaneProps & MainPanelProps) {
   const { openEditors, activeEditor: activeFile, activeEditor } = pane;
   const activeEditorAtoms = activeEditor?.contentAtoms;
+  console.log(MainPanelPane.name, activeEditor);
```
Comment (sehyod) [inline src/application/components/MainPanel.tsx:62]: I think this comes from debugging and can be removed :)

File: src/application/components/MainPanel.tsx [hunk 2]
```diff
@@ -97,17 +98,18 @@ interface MainPaneViewContentProps {
...
 
-  if (loadableFileContent.state === 'hasError') {
-    return <strong>Error: {String(loadableFileContent.error)}</strong>;
+  if (loadableEditorContent.state === 'hasError') {
+    return <strong>Error: {String(loadableEditorContent.error)}</strong>;
   }
 
-  const { data } = loadableFileContent;
+  const { data } = loadableEditorContent;
+  console.log(MainPaneViewContent.name, loadableEditorContent);
```
Comment (sehyod) [inline src/application/components/MainPanel.tsx:112]: Same here

File: src/features/references/editor/ReferencesTableView.tsx [hunk 3]
```diff
@@ -1,44 +1,165 @@
...
+
+  useEffect(() => setQuickFilter(defaultFilter), [defaultFilter]);
+
+  const [numberOfSelectedRows, setNumberOfSelectedRows] = useState(0);
+
+  const gridRef = useRef<AgGridReact<ReferenceItem>>(null);
+
+  const handleTitleEdit = useCallback(({ oldValue, newValue }: NewValueParams<ReferenceItem, 'string'>) => {
+    console.log('EDIT');
+    console.log(`${oldValue} -> ${newValue}`);
```
[the remark this answers is not in the document]
Comment (cguedes) [reply src/features/references/editor/ReferencesTableView.tsx:33]: These will be removed by the new implementation. Will keep them for now.

File: src/features/references/editor/grid/__tests__/mockLocalStorage.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  });
+
+  afterEach(() => {
+    setItemMock.mockRestore();
+    getItemMock.mockRestore();
+    removeItemMock.mockRestore();
+  });
+
+  return { setItemMock, getItemMock, removeItemMock };
+};
```
[the remark this answers is not in the document]
Comment (cguedes) [reply src/features/references/editor/grid/__tests__/mockLocalStorage.ts:19]: Sure yes.

**Review by sehyod — DISMISSED** [review]

**Review by sehyod — APPROVED** [review]

**Pull request merged**